File: wf-ui/lib/workflow/validate.ts
/**
 * Validation for workflow definitions. Port of gosample/internal/workflows/general/validate.go.
 * Uses ACTIVITY_NAMES from activityRegistry for activity name checks.
 */

import { isActivityRegistered } from "@/lib/activityRegistry";
import type { StepDef, ValueRef, WorkflowDef } from "@/lib/workflow/types";
import {
  MaxSignalTimeoutSeconds,
  MinSignalTimeoutSeconds,
  StepTypeActivity,
  StepTypeCondition,
  StepTypeSignal,
  StepTypeParallel,
  JoinAll,
  JoinAny,
  ValidJoinStrategies,
  SUPPORTED_DEFINITION_VERSIONS,
  VALID_CONDITION_OPERATORS,
  ValueRefLiteral,
  ValueRefInput,
  ValueRefStepResult,
} from "@/lib/workflow/types";

export interface ValidateOptions {
  /** If true, do not fail when some steps are unreachable (e.g. disconnected nodes). Use for export/copy. */
  allowUnreachable?: boolean;
}

export function validate(def: WorkflowDef | null, options?: ValidateOptions): string | null {
  if (def == null) return "workflow definition is nil";
  const steps = def.steps ?? {};
  const stepIds = Object.keys(steps);
  const allowUnreachable = options?.allowUnreachable === true;

  if (stepIds.length === 0) {
    if (def.startStepId != null && def.startStepId !== "")
      return `startStepId "${def.startStepId}" references non-existent step`;
    return null;
  }

  if (def.startStepId == null || def.startStepId === "")
    return "startStepId is required when there are steps";
  if (!(def.startStepId in steps))
    return `startStepId "${def.startStepId}" is not a key in steps`;

  if (def.version == null || def.version === "")
    return "definition version is required";
  if (!SUPPORTED_DEFINITION_VERSIONS[def.version])
    return `unsupported definition version "${def.version}"`;

  const referenced: Record<string, boolean> = { [def.startStepId]: true };
  for (const id of stepIds) {
    const err = validateStep(id, steps[id], steps, referenced);
    if (err) return err;
  }

  if (!allowUnreachable) {
    for (const id of stepIds) {
      if (!referenced[id])
        return `unreachable step "${id}" (not referenced from start or any step)`;
    }
  }

  const cycleErr = checkCycles(steps, def.startStepId, {});
  if (cycleErr) return cycleErr;
  return null;
}

function validateStep(
  id: string,
  step: StepDef,
  steps: Record<string, StepDef>,
  referenced: Record<string, boolean>
): string | null {
  switch (step.type) {
    case StepTypeActivity: {
      if (!step.name || step.name === "")
        return `step "${id}": activity step must have non-empty name`;
      if (!isActivityRegistered(step.name))
        return `step "${id}": unknown activity name "${step.name}"`;
      if (step.nextStepId != null && step.nextStepId !== "") {
        referenced[step.nextStepId] = true;
        if (!(step.nextStepId in steps))
          return `step "${id}": nextStepId "${step.nextStepId}" is not a key in steps`;
      }
      if (step.onFailureStepId != null && step.onFailureStepId !== "") {
        referenced[step.onFailureStepId] = true;
        if (!(step.onFailureStepId in steps))
          return `step "${id}": onFailureStepId "${step.onFailureStepId}" is not a key in steps`;
      }
      break;
    }
    case StepTypeSignal: {
      if (!step.signalName || step.signalName === "")
        return `step "${id}": signal step must have non-empty signalName`;
      if (
        step.signalTimeoutSeconds == null ||
        step.signalTimeoutSeconds < MinSignalTimeoutSeconds ||
        step.signalTimeoutSeconds > MaxSignalTimeoutSeconds
      )
        return `step "${id}": signalTimeoutSeconds must be between ${MinSignalTimeoutSeconds} and ${MaxSignalTimeoutSeconds}`;
      if (step.nextStepId == null || step.nextStepId === "")
        return `step "${id}": signal step must have nextStepId`;
      if (step.timeoutStepId == null || step.timeoutStepId === "")
        return `step "${id}": signal step must have timeoutStepId`;
      referenced[step.nextStepId] = true;
      if (!(step.nextStepId in steps))
        return `step "${id}": nextStepId "${step.nextStepId}" is not a key in steps`;
      referenced[step.timeoutStepId] = true;
      if (!(step.timeoutStepId in steps))
        return `step "${id}": timeoutStepId "${step.timeoutStepId}" is not a key in steps`;
      break;
    }
    case StepTypeCondition: {
      if (step.left == null)
        return `step "${id}": condition step must have left operand`;
      if (!step.operator || step.operator === "")
        return `step "${id}": condition step must have operator`;
      if (!VALID_CONDITION_OPERATORS[step.operator])
        return `step "${id}": invalid operator "${step.operator}"`;
      if (step.operator !== "exists" && step.operator !== "isEmpty") {
        if (step.right == null)
          return `step "${id}": condition with operator "${step.operator}" requires right operand`;
      }
      if (step.thenStepId != null && step.thenStepId !== "") {
        referenced[step.thenStepId] = true;
        if (!(step.thenStepId in steps))
          return `step "${id}": thenStepId "${step.thenStepId}" is not a key in steps`;
      }
      if (step.elseStepId != null && step.elseStepId !== "") {
        referenced[step.elseStepId] = true;
        if (!(step.elseStepId in steps))
          return `step "${id}": elseStepId "${step.elseStepId}" is not a key in steps`;
      }
      const leftErr = validateValueRef(step.left, steps, id, "left");
      if (leftErr) return leftErr;
      if (step.right != null) {
        const rightErr = validateValueRef(step.right, steps, id, "right");
        if (rightErr) return rightErr;
      }
      break;
    }
    case StepTypeParallel: {
      if (!step.branchStepIds || step.branchStepIds.length === 0)
        return `step "${id}": parallel step must have at least one branchStepId`;
      if (!step.join || step.join === "")
        return `step "${id}": parallel step must have join strategy (all or any)`;
      if (!ValidJoinStrategies[step.join])
        return `step "${id}": invalid join strategy "${step.join}" (must be all or any)`;
      // Validate each branch step exists and is not a parallel step (no nested parallel)
      for (const branchID of step.branchStepIds) {
        const branchStep = steps[branchID];
        if (!branchStep)
          return `step "${id}": branchStepId "${branchID}" is not a key in steps`;
        if (branchStep.type === StepTypeParallel)
          return `step "${id}": nested parallel steps are not supported (branchStepId "${branchID}")`;
        referenced[branchID] = true;
      }
      if (step.nextStepId != null && step.nextStepId !== "") {
        referenced[step.nextStepId] = true;
        if (!(step.nextStepId in steps))
          return `step "${id}": nextStepId "${step.nextStepId}" is not a key in steps`;
      }
      if (step.onFailureStepId != null && step.onFailureStepId !== "") {
        referenced[step.onFailureStepId] = true;
        if (!(step.onFailureStepId in steps))
          return `step "${id}": onFailureStepId "${step.onFailureStepId}" is not a key in steps`;
      }
      break;
    }
    default:
      return `step "${id}": invalid type "${step.type}" (must be activity, condition, signal, or parallel)`;
  }
  return null;
}

function validateValueRef(
  v: ValueRef | null | undefined,
  steps: Record<string, StepDef>,
  stepId: string,
  field: string
): string | null {
  if (v == null) return null;
  switch (v.type) {
    case ValueRefLiteral:
      break;
    case ValueRefInput:
      break;
    case ValueRefStepResult:
      if (!v.stepId || v.stepId === "")
        return `step "${stepId}": ${field} stepResult must have stepId`;
      if (!(v.stepId in steps))
        return `step "${stepId}": ${field} stepResult stepId "${v.stepId}" is not a key in steps`;
      break;
    default:
      return `step "${stepId}": ${field} has invalid value ref type "${v.type}"`;
  }
  return null;
}

function checkCycles(
  steps: Record<string, StepDef>,
  startId: string,
  path: Record<string, boolean>
): string | null {
  if (path[startId])
    return `cycle detected: step "${startId}" is reachable again`;
  const nextPath = { ...path, [startId]: true };
  const step = steps[startId];
  if (!step) return null;

  const nextIds: string[] = [];
  switch (step.type) {
    case StepTypeActivity:
      if (step.nextStepId != null && step.nextStepId !== "")
        nextIds.push(step.nextStepId);
      if (step.onFailureStepId != null && step.onFailureStepId !== "")
        nextIds.push(step.onFailureStepId);
      break;
    case StepTypeCondition:
      if (step.thenStepId != null && step.thenStepId !== "")
        nextIds.push(step.thenStepId);
      if (step.elseStepId != null && step.elseStepId !== "")
        nextIds.push(step.elseStepId);
      break;
    case StepTypeSignal:
      if (step.nextStepId != null && step.nextStepId !== "")
        nextIds.push(step.nextStepId);
      if (step.timeoutStepId != null && step.timeoutStepId !== "")
        nextIds.push(step.timeoutStepId);
      break;
    case StepTypeParallel:
      // For parallel steps, check cycles in each branch, then continue with nextStepId/onFailureStepId
      if (step.branchStepIds) {
        for (const branchID of step.branchStepIds) {
          const err = checkCycles(steps, branchID, nextPath);
          if (err) return err;
        }
      }
      if (step.nextStepId != null && step.nextStepId !== "")
        nextIds.push(step.nextStepId);
      if (step.onFailureStepId != null && step.onFailureStepId !== "")
        nextIds.push(step.onFailureStepId);
      break;
  }
  for (const nid of nextIds) {
    const err = checkCycles(steps, nid, nextPath);
    if (err) return err;
  }
  return null;
}

File: gosample/internal/workflows/general/validate.go
package general

import (
	"errors"
	"fmt"

	generalactivities "gosample/internal/activities/general"
)

// Validate checks the workflow definition: step refs, types, activity names, version, and no cycles.
func Validate(def *WorkflowDef) error {
	if def == nil {
		return errors.New("workflow definition is nil")
	}
	if len(def.Steps) == 0 {
		if def.StartStepID != nil && *def.StartStepID != "" {
			return fmt.Errorf("startStepId %q references non-existent step", *def.StartStepID)
		}
		return nil
	}

	// startStepId is required when there is at least one step (deterministic start, no map iteration)
	if def.StartStepID == nil || *def.StartStepID == "" {
		return errors.New("startStepId is required when there are steps")
	}
	if _, ok := def.Steps[*def.StartStepID]; !ok {
		return fmt.Errorf("startStepId %q is not a key in steps", *def.StartStepID)
	}

	// Version must be supported
	if def.Version == "" {
		return errors.New("definition version is required")
	}
	if !SupportedDefinitionVersions[def.Version] {
		return fmt.Errorf("unsupported definition version %q", def.Version)
	}

	// Validate each step and collect step IDs referenced by next/then/else/onFailure/timeout
	referenced := make(map[string]bool)
	referenced[*def.StartStepID] = true
	for id, step := range def.Steps {
		if err := validateStep(id, &step, def.Steps, referenced); err != nil {
			return err
		}
	}

	// Unreachable steps: every step must be start or referenced
	for id := range def.Steps {
		if !referenced[id] {
			return fmt.Errorf("unreachable step %q (not referenced from start or any step)", id)
		}
	}

	// Cycle check from definition start (resume start is validated in workflow when building input)
	if err := checkCycles(def.Steps, *def.StartStepID, nil); err != nil {
		return err
	}
	return nil
}

// CheckCycles verifies there are no cycles from startID. Call from workflow after resolving start (including resume).
func CheckCycles(steps map[string]StepDef, startID string) error {
	return checkCycles(steps, startID, nil)
}

func validateStep(id string, step *StepDef, steps map[string]StepDef, referenced map[string]bool) error {
	switch step.Type {
	case StepTypeActivity:
		if step.Name == "" {
			return fmt.Errorf("step %q: activity step must have non-empty name", id)
		}
		if !generalactivities.IsActivityRegistered(step.Name) {
			return fmt.Errorf("step %q: unknown activity name %q", id, step.Name)
		}
		if step.NextStepID != nil && *step.NextStepID != "" {
			referenced[*step.NextStepID] = true
			if _, ok := steps[*step.NextStepID]; !ok {
				return fmt.Errorf("step %q: nextStepId %q is not a key in steps", id, *step.NextStepID)
			}
		}
		if step.OnFailureStepID != nil && *step.OnFailureStepID != "" {
			referenced[*step.OnFailureStepID] = true
			if _, ok := steps[*step.OnFailureStepID]; !ok {
				return fmt.Errorf("step %q: onFailureStepId %q is not a key in steps", id, *step.OnFailureStepID)
			}
		}
	case StepTypeSignal:
		if step.SignalName == "" {
			return fmt.Errorf("step %q: signal step must have non-empty signalName", id)
		}
		if step.SignalTimeoutSeconds < MinSignalTimeoutSeconds || step.SignalTimeoutSeconds > MaxSignalTimeoutSeconds {
			return fmt.Errorf("step %q: signalTimeoutSeconds must be between %d and %d", id, MinSignalTimeoutSeconds, MaxSignalTimeoutSeconds)
		}
		if step.NextStepID == nil || *step.NextStepID == "" {
			return fmt.Errorf("step %q: signal step must have nextStepId", id)
		}
		if step.TimeoutStepID == nil || *step.TimeoutStepID == "" {
			return fmt.Errorf("step %q: signal step must have timeoutStepId", id)
		}
		referenced[*step.NextStepID] = true
		if _, ok := steps[*step.NextStepID]; !ok {
			return fmt.Errorf("step %q: nextStepId %q is not a key in steps", id, *step.NextStepID)
		}
		referenced[*step.TimeoutStepID] = true
		if _, ok := steps[*step.TimeoutStepID]; !ok {
			return fmt.Errorf("step %q: timeoutStepId %q is not a key in steps", id, *step.TimeoutStepID)
		}
	case StepTypeCondition:
		if step.Left == nil {
			return fmt.Errorf("step %q: condition step must have left operand", id)
		}
		if step.Operator == "" {
			return fmt.Errorf("step %q: condition step must have operator", id)
		}
		if !ValidConditionOperators[step.Operator] {
			return fmt.Errorf("step %q: invalid operator %q", id, step.Operator)
		}
		if step.Operator != OpExists && step.Operator != OpIsEmpty && step.Right == nil {
			return fmt.Errorf("step %q: condition with operator %q requires right operand", id, step.Operator)
		}
		if step.ThenStepID != nil && *step.ThenStepID != "" {
			referenced[*step.ThenStepID] = true
			if _, ok := steps[*step.ThenStepID]; !ok {
				return fmt.Errorf("step %q: thenStepId %q is not a key in steps", id, *step.ThenStepID)
			}
		}
		if step.ElseStepID != nil && *step.ElseStepID != "" {
			referenced[*step.ElseStepID] = true
			if _, ok := steps[*step.ElseStepID]; !ok {
				return fmt.Errorf("step %q: elseStepId %q is not a key in steps", id, *step.ElseStepID)
			}
		}
		if err := validateValueRef(step.Left, steps, id, "left"); err != nil {
			return err
		}
		if step.Right != nil {
			if err := validateValueRef(step.Right, steps, id, "right"); err != nil {
				return err
			}
		}
	case StepTypeParallel:
		if len(step.BranchStepIDs) == 0 {
			return fmt.Errorf("step %q: parallel step must have at least one branchStepId", id)
		}
		if step.Join == "" {
			return fmt.Errorf("step %q: parallel step must have join strategy (all or any)", id)
		}
		if !ValidJoinStrategies[step.Join] {
			return fmt.Errorf("step %q: invalid join strategy %q (must be all or any)", id, step.Join)
		}
		// Validate each branch step exists and is not a parallel step (no nested parallel)
		for _, branchID := range step.BranchStepIDs {
			branchStep, ok := steps[branchID]
			if !ok {
				return fmt.Errorf("step %q: branchStepId %q is not a key in steps", id, branchID)
			}
			if branchStep.Type == StepTypeParallel {
				return fmt.Errorf("step %q: nested parallel steps are not supported (branchStepId %q)", id, branchID)
			}
			referenced[branchID] = true
		}
		if step.NextStepID != nil && *step.NextStepID != "" {
			referenced[*step.NextStepID] = true
			if _, ok := steps[*step.NextStepID]; !ok {
				return fmt.Errorf("step %q: nextStepId %q is not a key in steps", id, *step.NextStepID)
			}
		}
		if step.OnFailureStepID != nil && *step.OnFailureStepID != "" {
			referenced[*step.OnFailureStepID] = true
			if _, ok := steps[*step.OnFailureStepID]; !ok {
				return fmt.Errorf("step %q: onFailureStepId %q is not a key in steps", id, *step.OnFailureStepID)
			}
		}
	default:
		return fmt.Errorf("step %q: invalid type %q (must be activity, condition, signal, or parallel)", id, step.Type)
	}
	return nil
}

func validateValueRef(v *ValueRef, steps map[string]StepDef, stepID, field string) error {
	if v == nil {
		return nil
	}
	switch v.Type {
	case ValueRefLiteral:
		// value can be anything
	case ValueRefInput:
		// path optional
	case ValueRefStepResult:
		if v.StepID == "" {
			return fmt.Errorf("step %q: %s stepResult must have stepId", stepID, field)
		}
		if _, ok := steps[v.StepID]; !ok {
			return fmt.Errorf("step %q: %s stepResult stepId %q is not a key in steps", stepID, field, v.StepID)
		}
	default:
		return fmt.Errorf("step %q: %s has invalid value ref type %q", stepID, field, v.Type)
	}
	return nil
}

// checkCycles does a DFS from start and returns an error if a cycle is found.
func checkCycles(steps map[string]StepDef, startID string, path map[string]bool) error {
	if path == nil {
		path = make(map[string]bool)
	}
	if path[startID] {
		return fmt.Errorf("cycle detected: step %q is reachable again", startID)
	}
	path[startID] = true
	defer func() { path[startID] = false }()

	step, ok := steps[startID]
	if !ok {
		return nil
	}

	var nextIDs []string
	switch step.Type {
	case StepTypeActivity:
		if step.NextStepID != nil && *step.NextStepID != "" {
			nextIDs = append(nextIDs, *step.NextStepID)
		}
		if step.OnFailureStepID != nil && *step.OnFailureStepID != "" {
			nextIDs = append(nextIDs, *step.OnFailureStepID)
		}
	case StepTypeCondition:
		if step.ThenStepID != nil && *step.ThenStepID != "" {
			nextIDs = append(nextIDs, *step.ThenStepID)
		}
		if step.ElseStepID != nil && *step.ElseStepID != "" {
			nextIDs = append(nextIDs, *step.ElseStepID)
		}
	case StepTypeSignal:
		if step.NextStepID != nil && *step.NextStepID != "" {
			nextIDs = append(nextIDs, *step.NextStepID)
		}
		if step.TimeoutStepID != nil && *step.TimeoutStepID != "" {
			nextIDs = append(nextIDs, *step.TimeoutStepID)
		}
	case StepTypeParallel:
		// For parallel steps, check cycles in each branch, then continue with nextStepId/onFailureStepId
		for _, branchID := range step.BranchStepIDs {
			if err := checkCycles(steps, branchID, path); err != nil {
				return err
			}
		}
		if step.NextStepID != nil && *step.NextStepID != "" {
			nextIDs = append(nextIDs, *step.NextStepID)
		}
		if step.OnFailureStepID != nil && *step.OnFailureStepID != "" {
			nextIDs = append(nextIDs, *step.OnFailureStepID)
		}
	}

	for _, nid := range nextIDs {
		if err := checkCycles(steps, nid, path); err != nil {
			return err
		}
	}
	return nil
}
